File: sdks/rust/src/storage.rs
//! Storage client — bindings for `/storage/v1/*` (ADR 0021 surface).
//!
//! Routes verified against `crates/basin-rest/src/server.rs:435-490`
//! (handlers in `routes/storage.rs` and `storage_sign.rs`):
//!
//! - `POST   /storage/v1/bucket`                                  create bucket
//! - `GET    /storage/v1/bucket/:name`                            bucket metadata
//! - `DELETE /storage/v1/bucket/:name`                            delete + purge
//! - `POST   /storage/v1/object/:bucket/*path`                    upload (JWT)
//! - `GET    /storage/v1/object/:bucket/*path`                    download (JWT)
//! - `DELETE /storage/v1/object/:bucket/*path`                    delete (JWT)
//! - `POST   /storage/v1/object/list/:bucket`                     list
//! - `DELETE /storage/v1/object/:bucket`                          bulk delete
//! - `GET    /storage/v1/object/public/:project_id/:bucket/*path` public dl
//! - `POST   /storage/v1/object/sign/upload/:bucket/*path`        mint signed URL
//! - `GET    /storage/v1/object/sign/:project/:bucket/*path`      signed download

use reqwest::Method;
use serde_json::{json, Value};

use crate::error::BasinError;
use crate::http::{encode_object_path, urlencode, Transport};
use crate::types::{Bucket, SignedUrl, StorageObject};

/// A downloaded object with its bytes and a couple of useful headers.
#[derive(Debug, Clone)]
pub struct DownloadResult {
    /// Raw object bytes.
    pub data: Vec<u8>,
    /// `Content-Type` header, if present.
    pub content_type: Option<String>,
    /// `ETag` header, if present.
    pub etag: Option<String>,
}

/// Storage client for bucket-level operations. Scope to a bucket with
/// [`from_bucket`](Self::from_bucket).
#[derive(Clone)]
pub struct StorageClient {
    transport: Transport,
    project_id: Option<String>,
}

impl StorageClient {
    pub(crate) fn new(transport: Transport, project_id: Option<String>) -> Self {
        Self {
            transport,
            project_id,
        }
    }

    /// `POST /storage/v1/bucket` (201).
    pub async fn create_bucket(
        &self,
        name: &str,
        public: bool,
        file_size_limit: Option<i64>,
        allowed_mime_types: &[String],
    ) -> Result<Bucket, BasinError> {
        let body = json!({
            "name": name,
            "public": public,
            "file_size_limit": file_size_limit,
            "allowed_mime_types": allowed_mime_types,
        });
        self.transport
            .request_json(Method::POST, "/storage/v1/bucket", &[], Some(&body), true)
            .await
    }

    /// `GET /storage/v1/bucket/:name`.
    pub async fn get_bucket(&self, name: &str) -> Result<Bucket, BasinError> {
        let path = format!("/storage/v1/bucket/{}", urlencode(name));
        self.transport
            .request_json(Method::GET, &path, &[], None, true)
            .await
    }

    /// `DELETE /storage/v1/bucket/:name` — also purges the object bytes.
    pub async fn delete_bucket(&self, name: &str) -> Result<(), BasinError> {
        let path = format!("/storage/v1/bucket/{}", urlencode(name));
        self.transport
            .request_discard(Method::DELETE, &path, &[], None, true)
            .await
    }

    /// Return an object-level client scoped to one bucket.
    pub fn from_bucket(&self, bucket: &str) -> StorageBucketClient {
        StorageBucketClient {
            transport: self.transport.clone(),
            bucket: bucket.to_string(),
            project_id: self.project_id.clone(),
        }
    }
}

/// Object-level storage operations scoped to a single bucket.
#[derive(Clone)]
pub struct StorageBucketClient {
    transport: Transport,
    bucket: String,
    project_id: Option<String>,
}

impl StorageBucketClient {
    fn object_path(&self, path: &str) -> String {
        format!(
            "/storage/v1/object/{}/{}",
            urlencode(&self.bucket),
            encode_object_path(path)
        )
    }

    /// `POST /storage/v1/object/:bucket/*path` — upload raw bytes.
    pub async fn upload(
        &self,
        path: &str,
        data: Vec<u8>,
        content_type: Option<&str>,
    ) -> Result<StorageObject, BasinError> {
        let mut headers = Vec::new();
        if let Some(ct) = content_type {
            headers.push(("content-type".to_string(), ct.to_string()));
        }
        let resp = self
            .transport
            .request(
                Method::POST,
                &self.object_path(path),
                &[],
                None,
                Some(data),
                &headers,
                true,
            )
            .await?;
        let text = resp.text().await.map_err(BasinError::from)?;
        serde_json::from_str(&text).map_err(|e| BasinError::Decode(e.to_string()))
    }

    /// `GET /storage/v1/object/:bucket/*path` — authenticated download.
    pub async fn download(&self, path: &str) -> Result<DownloadResult, BasinError> {
        let resp = self
            .transport
            .request(Method::GET, &self.object_path(path), &[], None, None, &[], true)
            .await?;
        let content_type = resp
            .headers()
            .get("content-type")
            .and_then(|v| v.to_str().ok())
            .map(str::to_string);
        let etag = resp
            .headers()
            .get("etag")
            .and_then(|v| v.to_str().ok())
            .map(str::to_string);
        let data = resp.bytes().await.map_err(BasinError::from)?.to_vec();
        Ok(DownloadResult {
            data,
            content_type,
            etag,
        })
    }

    /// `DELETE /storage/v1/object/:bucket/*path` — single delete.
    pub async fn remove(&self, path: &str) -> Result<(), BasinError> {
        self.transport
            .request_discard(Method::DELETE, &self.object_path(path), &[], None, true)
            .await
    }

    /// `POST /storage/v1/object/list/:bucket`.
    pub async fn list(
        &self,
        prefix: Option<&str>,
        limit: Option<i64>,
        offset: Option<i64>,
    ) -> Result<Vec<StorageObject>, BasinError> {
        let body = json!({ "prefix": prefix, "limit": limit, "offset": offset });
        let path = format!("/storage/v1/object/list/{}", urlencode(&self.bucket));
        Ok(self
            .transport
            .request_json_opt(Method::POST, &path, &[], Some(&body), true)
            .await?
            .unwrap_or_default())
    }

    /// `DELETE /storage/v1/object/:bucket` with `{ prefixes: [...] }`.
    ///
    /// Bulk-deletes every object matching any prefix. RLS-denied objects are
    /// skipped server-side.
    pub async fn remove_by_prefixes(&self, prefixes: &[String]) -> Result<(), BasinError> {
        let body = json!({ "prefixes": prefixes });
        let path = format!("/storage/v1/object/{}", urlencode(&self.bucket));
        self.transport
            .request_discard(Method::DELETE, &path, &[], Some(&body), true)
            .await
    }

    /// Build the no-JWT public URL.
    ///
    /// `GET /storage/v1/object/public/:project_id/:bucket/*path`. Only works
    /// for buckets with `public = true`. Errors if no project id is known.
    pub fn get_public_url(&self, path: &str) -> Result<String, BasinError> {
        let project = self.project_id.as_ref().ok_or_else(|| {
            BasinError::invalid_request(
                "public URLs require a project id (set project_id on the client builder)",
            )
        })?;
        Ok(format!(
            "{}/storage/v1/object/public/{}/{}/{}",
            self.transport.base_url,
            urlencode(project),
            urlencode(&self.bucket),
            encode_object_path(path)
        ))
    }

    /// `POST /storage/v1/object/sign/upload/:bucket/*path` — mint a signed URL.
    ///
    /// TTL is capped at 7 days server-side; default `3600` s. Despite the
    /// `upload` literal in the path this mints a **download** URL — that
    /// segment only disambiguates the axum router (see `storage_sign.rs`).
    pub async fn create_signed_url(
        &self,
        path: &str,
        expires_in: i64,
    ) -> Result<SignedUrl, BasinError> {
        let body = json!({ "expires_in": expires_in });
        let req_path = format!(
            "/storage/v1/object/sign/upload/{}/{}",
            urlencode(&self.bucket),
            encode_object_path(path)
        );
        let data: Value = self
            .transport
            .request_json(Method::POST, &req_path, &[], Some(&body), true)
            .await?;
        Ok(SignedUrl::from_value(&data, &self.transport.base_url))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use reqwest::Client;

    fn bucket() -> StorageBucketClient {
        StorageBucketClient {
            transport: Transport::new("http://h".into(), None, Client::new()),
            bucket: "media".into(),
            project_id: Some("01PROJ".into()),
        }
    }

    #[test]
    fn object_path_encodes_segments() {
        let b = bucket();
        assert_eq!(
            b.object_path("a b/c.png"),
            "/storage/v1/object/media/a%20b/c.png"
        );
    }

    #[test]
    fn public_url_built_with_project() {
        let b = bucket();
        assert_eq!(
            b.get_public_url("x/y.png").unwrap(),
            "http://h/storage/v1/object/public/01PROJ/media/x/y.png"
        );
    }

    #[test]
    fn public_url_requires_project() {
        let mut b = bucket();
        b.project_id = None;
        assert!(b.get_public_url("x").is_err());
    }
}

File: sdks/go/storage.go
package basin

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
)

// StorageClient provides bucket management for /storage/v1/bucket/*.
//
// Route sources (verified, crates/basin-rest/src/server.rs:373-424):
//
//	POST   /storage/v1/bucket                    create bucket
//	GET    /storage/v1/bucket/:name              bucket metadata
//	DELETE /storage/v1/bucket/:name              delete + purge
type StorageClient struct {
	t          *transport
	projectFn  func() string
}

func newStorageClient(t *transport, projectFn func() string) *StorageClient {
	return &StorageClient{t: t, projectFn: projectFn}
}

// CreateBucket creates a new storage bucket.
// POST /storage/v1/bucket (201).
func (s *StorageClient) CreateBucket(ctx context.Context, name string, opts ...BucketOption) (*Bucket, error) {
	cfg := &bucketConfig{AllowedMimeTypes: []string{}}
	for _, o := range opts {
		o(cfg)
	}
	body := map[string]any{
		"name":               name,
		"public":             cfg.Public,
		"allowed_mime_types": cfg.AllowedMimeTypes,
	}
	if cfg.FileSizeLimit != nil {
		body["file_size_limit"] = *cfg.FileSizeLimit
	}
	raw, _, err := s.t.doJSON(ctx, "POST", "/storage/v1/bucket", nil, body, true)
	if err != nil {
		return nil, err
	}
	var bucket Bucket
	if err := json.Unmarshal(raw, &bucket); err != nil {
		return nil, fmt.Errorf("basin: decode bucket: %w", err)
	}
	return &bucket, nil
}

// GetBucket retrieves bucket metadata.
// GET /storage/v1/bucket/:name.
func (s *StorageClient) GetBucket(ctx context.Context, name string) (*Bucket, error) {
	raw, _, err := s.t.doJSON(ctx, "GET", "/storage/v1/bucket/"+urlEncodeSegment(name), nil, nil, true)
	if err != nil {
		return nil, err
	}
	var bucket Bucket
	if err := json.Unmarshal(raw, &bucket); err != nil {
		return nil, fmt.Errorf("basin: decode bucket: %w", err)
	}
	return &bucket, nil
}

// DeleteBucket deletes a bucket and purges its object bytes.
// DELETE /storage/v1/bucket/:name.
func (s *StorageClient) DeleteBucket(ctx context.Context, name string) error {
	_, _, err := s.t.doJSON(ctx, "DELETE", "/storage/v1/bucket/"+urlEncodeSegment(name), nil, nil, true)
	return err
}

// FromBucket returns an object-level client scoped to one bucket.
func (s *StorageClient) FromBucket(bucket string) *StorageBucketClient {
	return &StorageBucketClient{t: s.t, bucket: bucket, projectFn: s.projectFn}
}

// BucketOption configures bucket creation.
type BucketOption func(*bucketConfig)

type bucketConfig struct {
	Public           bool
	FileSizeLimit    *int64
	AllowedMimeTypes []string
}

// WithPublic makes the bucket publicly accessible (no JWT required for downloads).
func WithPublic(public bool) BucketOption {
	return func(c *bucketConfig) { c.Public = public }
}

// WithFileSizeLimit sets a per-object byte limit. 0 means no limit.
func WithFileSizeLimit(n int64) BucketOption {
	return func(c *bucketConfig) { c.FileSizeLimit = &n }
}

// WithAllowedMimeTypes restricts the MIME types accepted on upload.
func WithAllowedMimeTypes(types []string) BucketOption {
	return func(c *bucketConfig) { c.AllowedMimeTypes = types }
}

// ---------------------------------------------------------------------------
// StorageBucketClient — object operations
// ---------------------------------------------------------------------------

// StorageBucketClient provides object-level operations scoped to one bucket.
//
// Route sources (verified, crates/basin-rest/src/server.rs):
//
//	POST   /storage/v1/object/:bucket/*path                    upload
//	GET    /storage/v1/object/:bucket/*path                    download
//	DELETE /storage/v1/object/:bucket/*path                    single delete
//	POST   /storage/v1/object/list/:bucket                     list
//	DELETE /storage/v1/object/:bucket                          bulk delete
//	GET    /storage/v1/object/public/:project_id/:bucket/*path public download (no JWT)
//	POST   /storage/v1/object/sign/upload/:bucket/*path        mint signed URL
//	GET    /storage/v1/object/sign/:project/:bucket/*path      signed download (no JWT)
type StorageBucketClient struct {
	t          *transport
	bucket     string
	projectFn  func() string
}

// UploadOption configures object uploads.
type UploadOption func(*uploadConfig)

type uploadConfig struct {
	ContentType string
}

// WithContentType sets the Content-Type header for the upload.
func WithContentType(ct string) UploadOption {
	return func(c *uploadConfig) { c.ContentType = ct }
}

// Upload posts raw bytes to /storage/v1/object/:bucket/*path.
func (b *StorageBucketClient) Upload(ctx context.Context, path string, data []byte, opts ...UploadOption) (*StorageObject, error) {
	cfg := &uploadConfig{}
	for _, o := range opts {
		o(cfg)
	}
	objPath := b.objectPath(path)
	raw, _, err := b.t.doUpload(ctx, "POST", objPath, data, cfg.ContentType, true)
	if err != nil {
		return nil, err
	}
	if raw == nil {
		return &StorageObject{}, nil
	}
	var obj StorageObject
	if err := json.Unmarshal(raw, &obj); err != nil {
		return nil, fmt.Errorf("basin: decode storage object: %w", err)
	}
	return &obj, nil
}

// Download fetches /storage/v1/object/:bucket/*path (authenticated).
func (b *StorageBucketClient) Download(ctx context.Context, path string) (*DownloadResult, error) {
	resp, err := b.t.doRaw(ctx, "GET", b.objectPath(path), nil, nil, true)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		body, _ := io.ReadAll(resp.Body)
		return nil, b.t.parseError(body, resp.StatusCode)
	}

	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("basin: read download body: %w", err)
	}
	return &DownloadResult{
		Data:        data,
		ContentType: resp.Header.Get("Content-Type"),
		ETag:        resp.Header.Get("ETag"),
	}, nil
}

// Remove deletes a single object.
// DELETE /storage/v1/object/:bucket/*path.
func (b *StorageBucketClient) Remove(ctx context.Context, path string) error {
	_, _, err := b.t.doJSON(ctx, "DELETE", b.objectPath(path), nil, nil, true)
	return err
}

// List returns objects in the bucket matching prefix.
// POST /storage/v1/object/list/:bucket (body: { prefix, limit, offset }).
func (b *StorageBucketClient) List(ctx context.Context, prefix string, limit, offset int) ([]StorageObject, error) {
	body := map[string]any{
		"prefix": prefix,
	}
	if limit > 0 {
		body["limit"] = limit
	}
	if offset > 0 {
		body["offset"] = offset
	}
	raw, _, err := b.t.doJSON(ctx, "POST",
		"/storage/v1/object/list/"+urlEncodeSegment(b.bucket), nil, body, true)
	if err != nil {
		return nil, err
	}
	if raw == nil {
		return []StorageObject{}, nil
	}
	var objects []StorageObject
	if err := json.Unmarshal(raw, &objects); err != nil {
		return nil, fmt.Errorf("basin: decode object list: %w", err)
	}
	return objects, nil
}

// RemoveByPrefixes bulk-deletes all objects matching any prefix.
// DELETE /storage/v1/object/:bucket with { prefixes: [...] }.
func (b *StorageBucketClient) RemoveByPrefixes(ctx context.Context, prefixes []string) error {
	_, _, err := b.t.doJSON(ctx, "DELETE",
		"/storage/v1/object/"+urlEncodeSegment(b.bucket), nil,
		map[string]any{"prefixes": prefixes}, true)
	return err
}

// GetPublicURL returns the no-JWT public download URL for an object.
// Only works for buckets with Public=true.
// GET /storage/v1/object/public/:project_id/:bucket/*path.
func (b *StorageBucketClient) GetPublicURL(path string) (string, error) {
	var project string
	if b.projectFn != nil {
		project = b.projectFn()
	}
	if project == "" {
		return "", newBasinError("E_INVALID_REQUEST",
			"public URLs require a project id (pass WithProjectID to New())", 0)
	}
	return b.t.baseURL + "/storage/v1/object/public/" + project +
		"/" + urlEncodeSegment(b.bucket) + "/" + encodeObjectPath(path), nil
}

// CreateSignedURL mints a time-limited signed download URL.
// POST /storage/v1/object/sign/upload/:bucket/*path.
// TTL is capped at 7 days server-side; default 3600s.
// Despite the "upload" literal in the path, this mints a download URL —
// that segment only disambiguates the axum router (storage_sign.rs).
func (b *StorageBucketClient) CreateSignedURL(ctx context.Context, path string, expiresIn int) (*SignedURL, error) {
	if expiresIn <= 0 {
		expiresIn = 3600
	}
	mintPath := "/storage/v1/object/sign/upload/" +
		urlEncodeSegment(b.bucket) + "/" + encodeObjectPath(path)
	raw, _, err := b.t.doJSON(ctx, "POST", mintPath, nil, map[string]any{
		"expires_in": expiresIn,
	}, true)
	if err != nil {
		return nil, err
	}
	var signed SignedURL
	if err := json.Unmarshal(raw, &signed); err != nil {
		return nil, fmt.Errorf("basin: decode signed url: %w", err)
	}
	// Build the full URL — server returns a relative path in signedUrl.
	signed.AbsoluteURL = b.t.baseURL + signed.SignedURL
	return &signed, nil
}

// DownloadSigned fetches a signed URL directly (no JWT required).
// The signedPath is the value from SignedURL.SignedURL (server-relative).
func (b *StorageBucketClient) DownloadSigned(ctx context.Context, signedPath string) (*DownloadResult, error) {
	rawURL := b.t.baseURL + signedPath
	req, err := http.NewRequestWithContext(ctx, "GET", rawURL, nil)
	if err != nil {
		return nil, fmt.Errorf("basin: build request: %w", err)
	}
	resp, err := b.t.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("basin: http: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		body, _ := io.ReadAll(resp.Body)
		return nil, b.t.parseError(body, resp.StatusCode)
	}
	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("basin: read download body: %w", err)
	}
	return &DownloadResult{
		Data:        data,
		ContentType: resp.Header.Get("Content-Type"),
		ETag:        resp.Header.Get("ETag"),
	}, nil
}

// ---------------------------------------------------------------------------
// Path helpers
// ---------------------------------------------------------------------------

func (b *StorageBucketClient) objectPath(path string) string {
	return "/storage/v1/object/" + urlEncodeSegment(b.bucket) + "/" + encodeObjectPath(path)
}

// encodeObjectPath percent-encodes each path segment while preserving slashes.
func encodeObjectPath(path string) string {
	segments := strings.Split(path, "/")
	encoded := make([]string, len(segments))
	for i, seg := range segments {
		encoded[i] = url.PathEscape(seg)
	}
	return strings.Join(encoded, "/")
}

// urlEncodeSegment percent-encodes a single path segment (no slashes).
func urlEncodeSegment(s string) string {
	return url.PathEscape(s)
}
